lib/lightning/usage_tracking/workflow_metrics_service.ex

```diff
@@ -4,25 +4,79 @@ defmodule Lightning.UsageTracking.WorkflowMetricsService do
 
 
   """
+
+  alias Lightning.Repo
   alias Lightning.UsageTracking.RunService
+  alias Lightning.UsageTracking.WorkflowMetricsQuery
 
   def find_eligible_workflows(workflows, date) do
     workflows
     |> Enum.filter(fn workflow -> eligible_workflow?(workflow, date) end)
   end
 
   def generate_metrics(workflow, cleartext_enabled, date) do
-    runs = RunService.finished_runs(workflow.runs, date)
-    steps = RunService.finished_steps(workflow.runs, date)
-    active_jobs = RunService.unique_job_ids(steps, date)
+    workflow
+    |> metrics_for_runs(date)
+    |> add_active_jobs()
+    |> Map.delete(:job_ids)
+    |> Map.merge(instrument_identity(workflow.id, cleartext_enabled))
+  end
+
+  defp metrics_for_runs(workflow, date) do
+    chunk_size = Lightning.Config.usage_tracking_run_chunk_size()
+
+    {:ok, updated_metrics} =
+      Repo.transaction(fn ->
+        workflow
+        |> WorkflowMetricsQuery.workflow_runs()
+        |> WorkflowMetricsQuery.runs_finished_on(date)
+        |> Repo.stream(max_rows: chunk_size)
+        |> Stream.chunk_every(chunk_size)
+        |> Enum.flat_map(&preload_assocs/1)
+        |> Enum.reduce(base_metrics(workflow), metric_updater(date))
+      end)
+
+    updated_metrics
+  end
+
+  defp preload_assocs(run_chunk), do: Repo.preload(run_chunk, steps: [:job])
 
+  defp base_metrics(%{jobs: jobs}) do
     %{
-      no_of_active_jobs: Enum.count(active_jobs),
-      no_of_jobs: Enum.count(workflow.jobs),
-      no_of_runs: Enum.count(runs),
-      no_of_steps: Enum.count(steps)
+      no_of_jobs: Enum.count(jobs),
+      no_of_runs: 0,
+      no_of_steps: 0,
+      job_ids: []
     }
-    |> Map.merge(instrument_identity(workflow.id, cleartext_enabled))
+  end
+
+  defp metric_updater(date) do
+    fn run, acc ->
+      %{
+        no_of_steps: no_of_steps,
+        no_of_runs: no_of_runs,
+        job_ids: job_ids
+      } = acc
+
+      steps = RunService.finished_steps(run, date)
+      active_jobs = RunService.unique_job_ids(steps, date)
+
+      Map.merge(
+        acc,
+        %{
+          no_of_runs: no_of_runs + 1,
+          no_of_steps: no_of_steps + Enum.count(steps),
+          job_ids: Enum.concat(job_ids, active_jobs)
+        }
+      )
```

Comment:
While this is similar to the original implementation I'm curious if all of this could be rolled into a query as well. That way we could skip the cost of hydrating steps as well. But lets roll with this for now, the preload is much much smaller and with the chunking we've probably cut this down significantly.

Reply:
@stuartc I did consider this as a follow-up PR, but I stopped because I think a single query rolling up all the things is much harder to debug. For example, if there is an issue with some of the data going through, I would find it easier to pull this approach apart in an IEx session than I would a single query.

So, I held off on the suggestion until I have more data that indicates that this optimisation (i.e. a smarter query) is needed.